docs/dev/internals/_stimulus-backend.md

```diff
@@ -0,0 +1,173 @@
+---
+title: "Turbo-compatible Stimulus controllers"
+description: "Howto write Stimulus controllers for a Turbo application (that don't suck)"
+---
+
+**Lifecycle callback cheatsheet**
+
+```js
+import {Controller} from '@hotwired/stimulus';
+
+export default class extends Controller {
+    // Run once, when this controller class is registered (rarely needed)
+    static afterLoad(identifier, application) {
+    }
+
+    // Run once, when the current controller instance is created
+    initialize() {
+    }
+
+    // Run each time the controller element is added to the DOM
+    connect() {
+    }
+
+    // Run each time the controller element is removed from the DOM
+    disconnect() {
+    }
+
+    // Run before Turbo creates a cache snapshot (Contao-only)
+    beforeCache(event) {
+    }
+}
+
+```
+
+## 1) Use idempotent transformations
+
+If you need to make changes to the DOM make sure these are idempotent. That means, applying the code multiple times does
+not do any harm. Why is this important? Because cache entries are made before the DOM gets removed, so any cleanup in a
+`disconnect()` method has no effect.
```

Comment:
I am a bit confused here. You are saying that any cleanup in a `disconnect()` method has no effect, but further down in 2) you are also saying:

> Cleanup CSS classes on parent elements, created sibling elements, etc.

So shouldn't it be enough to delete and created sibling elements in the `disconnect()` method rather than having to rely on the `turbo:before-cache` event?

Reply:
The order for a page transition with caching is `beforeCache`, then `disconnect`. So from the perspective of the snapshot the disconnect has no effect. 

The `disconnect` is still crucial for when the DOM/parts of it get exchanged. Memory cleanup, removing parent classes etc. This will happen without `beforeCache` calls and even on the cached snapshot when it is displayed as a preview before the actual content has arrived.

Reply:
Basically: `beforeCache` must shape the DOM, so that a `connect` can run on it, `disconnect` must restore the DOM to its original state.

Reply:
Feel free to improve the text, if we can make this more clear.